File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.java
class Solution {
    
    public int size(ListNode head) {
        int s = 0;
        while (head != null) {
            s++;
            head = head.next;
        }
        return s;
    }
    
    public ListNode deleteMiddle(ListNode head) {
        int len = size(head), curr = 0;
        if (len == 1) return null;
        int middle = len / 2;
        ListNode current = head;
        while (curr != middle - 1) {
            current = current.next;
            curr++;
        }
        current.next = current.next.next;
        return head;
    }
}

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.cpp
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        if (head->next == nullptr) {
            delete head;
            return nullptr;
        }
        
        ListNode *slow = head, *fast = head, *trail = head;
        
        while (fast != nullptr && fast->next != nullptr) {
            trail = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        
        trail->next = slow->next;
        slow->next = nullptr;
        delete slow;
        
        return head;
    }
};